File: src/services/pricing.js
const { getRetailCatalog } = require("./catalog");

const SHIPPING_COSTS = {
  delivery: 100,
  cvs: 60,
  store: 0,
};

function createCheckoutError(message) {
  const error = new Error(message);
  error.statusCode = 400;
  return error;
}

function getFreeShippingThreshold(retailCatalog) {
  return (
    retailCatalog.section_meta?.shipping?.free_shipping_threshold_NTD || 1500
  );
}

function getShippingCost(method, subtotal, threshold) {
  const cost = SHIPPING_COSTS[method];

  if (typeof cost !== "number") {
    throw createCheckoutError("Invalid shipping method");
  }

  return subtotal >= threshold ? 0 : cost;
}

function toPublicProduct(product) {
  return {
    id: product.id,
    name_zh: product.name_zh,
    name_en: product.name_en,
    origin_zh: product.origin_zh,
    origin_en: product.origin_en,
    process_zh: product.process_zh,
    process_en: product.process_en,
    flavor_zh: product.flavor_zh,
    flavor_en: product.flavor_en,
    price_NTD: product.price_NTD,
    weight_g: product.weight_g,
    roast_level: product.roast_level,
    roast_level_zh: product.roast_level_zh,
    image: product.image,
    in_stock: Boolean(product.in_stock),
    variants: (product.variants || []).map((variant) => ({
      sku: variant.sku,
      grind: variant.grind,
      grind_zh: variant.grind_zh,
      grind_en: variant.grind_en,
      weight_g: variant.weight_g,
      price_NTD: variant.price_NTD,
      in_stock: Boolean(variant.in_stock),
    })),
  };
}

async function getPublicRetailCatalog() {
  const retailCatalog = await getRetailCatalog();

  return {
    items: retailCatalog.items.map(toPublicProduct),
    shipping: {
      freeShippingThreshold: getFreeShippingThreshold(retailCatalog),
      costs: SHIPPING_COSTS,
      methods_zh: retailCatalog.section_meta?.shipping?.methods_zh || [],
      methods_en: retailCatalog.section_meta?.shipping?.methods_en || [],
    },
  };
}

async function buildPricedOrder(requestedItems, shippingMethod) {
  const retailCatalog = await getRetailCatalog();
  const skuIndex = new Map();

  retailCatalog.items.forEach((product) => {
    (product.variants || []).forEach((variant) => {
      skuIndex.set(variant.sku, { product, variant });
    });
  });

  const items = requestedItems.map((requestedItem) => {
    const entry = skuIndex.get(requestedItem.sku);

    if (!entry) {
      throw createCheckoutError(
        "Cart contains an unknown product. Please refresh and try again.",
      );
    }

    const { product, variant } = entry;

    if (!product.in_stock || !variant.in_stock) {
      throw createCheckoutError(
        `${product.name_zh} is currently out of stock.`,
      );
    }

    const price = variant.price_NTD;
    const subtotal = price * requestedItem.qty;

    return {
      sku: variant.sku,
      productId: product.id,
      name: `${product.name_zh}（${variant.grind_zh}/${variant.weight_g}g）`,
      name_zh: product.name_zh,
      name_en: product.name_en,
      grind: variant.grind,
      grind_zh: variant.grind_zh,
      grind_en: variant.grind_en,
      weight: variant.weight_g,
      price,
      qty: requestedItem.qty,
      subtotal,
      ecpayName: product.ecpay?.ItemName || product.name_zh,
    };
  });

  const subtotal = items.reduce((sum, item) => sum + item.subtotal, 0);
  const freeShippingThreshold = getFreeShippingThreshold(retailCatalog);
  const shippingCost = getShippingCost(
    shippingMethod,
    subtotal,
    freeShippingThreshold,
  );
  const total = subtotal + shippingCost;
  const itemName = items
    .map(
      (item) => `${item.ecpayName}#${item.price}#${item.qty}#${item.subtotal}`,
    )
    .join("|")
    .slice(0, 400);

  return {
    items: items.map(({ ecpayName, ...item }) => item),
    itemName,
    pricing: {
      subtotal,
      shippingCost,
      total,
      freeShippingThreshold,
      currency: retailCatalog.section_meta?.currency || "NTD",
    },
  };
}

module.exports = {
  SHIPPING_COSTS,
  buildPricedOrder,
  getPublicRetailCatalog,
};

File: app/services/pricing.py
from app.services.catalog import get_retail_section, SHIPPING_COSTS


class CheckoutError(Exception):
    status_code = 400


def _free_shipping_threshold(retail):
    return retail.get("section_meta", {}).get("shipping", {}).get("free_shipping_threshold_NTD", 1500)


def _shipping_cost(method: str, subtotal: int, threshold: int) -> int:
    if method not in SHIPPING_COSTS:
        raise CheckoutError("Invalid shipping method")
    cost = SHIPPING_COSTS[method]
    return 0 if subtotal >= threshold else cost


def build_priced_order(requested_items: list, shipping_method: str) -> dict:
    retail = get_retail_section()

    # Build SKU index
    sku_index = {}
    for product in retail.get("items", []):
        for variant in product.get("variants", []):
            sku_index[variant["sku"]] = {"product": product, "variant": variant}

    items = []
    for req in requested_items:
        entry = sku_index.get(req["sku"])
        if not entry:
            raise CheckoutError("Cart contains an unknown product. Please refresh and try again.")

        product = entry["product"]
        variant = entry["variant"]

        if not product.get("in_stock") or not variant.get("in_stock"):
            raise CheckoutError(f"{product['name_zh']} is currently out of stock.")

        price = variant["price_NTD"]
        qty = req["qty"]
        subtotal = price * qty

        items.append({
            "sku": variant["sku"],
            "name": f"{product['name_zh']}（{variant['grind_zh']}/{variant['weight_g']}g）",
            "name_zh": product["name_zh"],
            "name_en": product["name_en"],
            "grind": variant["grind"],
            "grind_zh": variant["grind_zh"],
            "grind_en": variant["grind_en"],
            "weight": variant["weight_g"],
            "price": price,
            "qty": qty,
            "subtotal": subtotal,
            "ecpay_name": product.get("ecpay", {}).get("ItemName") or product["name_zh"],
        })

    subtotal_total = sum(i["subtotal"] for i in items)
    threshold = _free_shipping_threshold(retail)
    shipping_cost = _shipping_cost(shipping_method, subtotal_total, threshold)
    total = subtotal_total + shipping_cost

    item_name = "|".join(
        f"{i['ecpay_name']}#{i['price']}#{i['qty']}#{i['subtotal']}"
        for i in items
    )[:400]

    # Strip ecpay_name from returned items
    clean_items = [{k: v for k, v in i.items() if k != "ecpay_name"} for i in items]

    return {
        "items": clean_items,
        "item_name": item_name,
        "pricing": {
            "subtotal": subtotal_total,
            "shipping_cost": shipping_cost,
            "total": total,
            "free_shipping_threshold": threshold,
            "currency": retail.get("section_meta", {}).get("currency", "NTD"),
        },
    }
